Importing libraries

In [1]:
import re
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

Importing methods

In [2]:
from nltk.tokenize import sent_tokenize
from nltk.tokenize import word_tokenize

storing text from a news article

In [3]:
raw_text = ""

In [4]:
raw_text = "In this chapter we'll write a computer program implementing a neural network that learns to recognize handwritten digits. The program is just 74 lines long, and uses no special neural network libraries. But this short program can recognize digits with an accuracy over 96 percent, without human intervention. Furthermore, in later chapters we'll develop ideas which can improve accuracy to over 99 percent. In fact, the best commercial neural networks are now so good that they are used by banks to process cheques, and by post offices to recognize addresses. We're focusing on handwriting recognition because it's an excellent prototype problem for learning about neural networks in general. As a prototype it hits a sweet spot: it's challenging - it's no small feat to recognize handwritten digits - but it's not so difficult as to require an extremely complicated solution, or tremendous computational power. Furthermore, it's a great way to develop more advanced techniques, such as deep learning. And so throughout the book we'll return repeatedly to the problem of handwriting recognition. Later in the book, we'll discuss how these ideas may be applied to other problems in computer vision, and also in speech, natural language processing, and other domains. Of course, if the point of the chapter was only to write a computer program to recognize handwritten digits, then the chapter would be much shorter! But along the way we'll develop many key ideas about neural networks, including two important types of artificial neuron (the perceptron and the sigmoid neuron), and the standard learning algorithm for neural networks, known as stochastic gradient descent. Throughout, I focus on explaining why things are done the way they are, and on building your neural networks intuition. That requires a lengthier discussion than if I just presented the basic mechanics of what's going on, but it's worth it for the deeper understanding you'll attain. Amongst the payoffs, by the end of the chapter we'll be in position to understand what deep learning is, and why it matters."


Checking the length of the raw text we have.

In [5]:
len(raw_text)

2080

Now, we need to tokenize the text we have.

In [6]:
words = word_tokenize(raw_text)

In [7]:
len(words)

389

In [8]:
words

['In',
 'this',
 'chapter',
 'we',
 "'ll",
 'write',
 'a',
 'computer',
 'program',
 'implementing',
 'a',
 'neural',
 'network',
 'that',
 'learns',
 'to',
 'recognize',
 'handwritten',
 'digits',
 '.',
 'The',
 'program',
 'is',
 'just',
 '74',
 'lines',
 'long',
 ',',
 'and',
 'uses',
 'no',
 'special',
 'neural',
 'network',
 'libraries',
 '.',
 'But',
 'this',
 'short',
 'program',
 'can',
 'recognize',
 'digits',
 'with',
 'an',
 'accuracy',
 'over',
 '96',
 'percent',
 ',',
 'without',
 'human',
 'intervention',
 '.',
 'Furthermore',
 ',',
 'in',
 'later',
 'chapters',
 'we',
 "'ll",
 'develop',
 'ideas',
 'which',
 'can',
 'improve',
 'accuracy',
 'to',
 'over',
 '99',
 'percent',
 '.',
 'In',
 'fact',
 ',',
 'the',
 'best',
 'commercial',
 'neural',
 'networks',
 'are',
 'now',
 'so',
 'good',
 'that',
 'they',
 'are',
 'used',
 'by',
 'banks',
 'to',
 'process',
 'cheques',
 ',',
 'and',
 'by',
 'post',
 'offices',
 'to',
 'recognize',
 'addresses',
 '.',
 'We',
 "'re",
 'foc

We have 389 words with us. We need to remove the punctuations that were counted as a word by word tokenizer.

In [9]:
import string

In [10]:
only_words = [w for w in words if not w in string.punctuation]

In [11]:
print(len(only_words))
print(only_words)

347
['In', 'this', 'chapter', 'we', "'ll", 'write', 'a', 'computer', 'program', 'implementing', 'a', 'neural', 'network', 'that', 'learns', 'to', 'recognize', 'handwritten', 'digits', 'The', 'program', 'is', 'just', '74', 'lines', 'long', 'and', 'uses', 'no', 'special', 'neural', 'network', 'libraries', 'But', 'this', 'short', 'program', 'can', 'recognize', 'digits', 'with', 'an', 'accuracy', 'over', '96', 'percent', 'without', 'human', 'intervention', 'Furthermore', 'in', 'later', 'chapters', 'we', "'ll", 'develop', 'ideas', 'which', 'can', 'improve', 'accuracy', 'to', 'over', '99', 'percent', 'In', 'fact', 'the', 'best', 'commercial', 'neural', 'networks', 'are', 'now', 'so', 'good', 'that', 'they', 'are', 'used', 'by', 'banks', 'to', 'process', 'cheques', 'and', 'by', 'post', 'offices', 'to', 'recognize', 'addresses', 'We', "'re", 'focusing', 'on', 'handwriting', 'recognition', 'because', 'it', "'s", 'an', 'excellent', 'prototype', 'problem', 'for', 'learning', 'about', 'neural', 'n

We see that 42 'words' have been removed. Those were the punctuations.

In [12]:
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

Upon inspection, we saw that double apostrophes still remain. We need to explicitly remove them.

only_words.remove("“"); only_words.remove("”"); only_words.remove("–")

In [13]:
len(only_words)

347

In [14]:
import re
from nltk.corpus import wordnet
R_patterns = [
   (r'won\'t', 'will not'),
   (r'can\'t', 'cannot'),
   (r'i\'m', 'i am'),
   (r'(\w+)\'ll', '\g<1> will'),
   (r'(\w+)n\'t', '\g<1> not'),
   (r'(\w+)\'ve', '\g<1> have'),
   (r'(\w+)\'s', '\g<1> is'),
   (r'(\w+)\'re', '\g<1> are'),
]

Now we need to convert words

In [15]:
pattern = [(re.compile(regex), repl) for (regex, repl) in R_patterns ]

In [16]:
s = raw_text
len(s)

2080

In [17]:
#s = raw_text
for (pat, repl) in pattern:
    s = re.sub(pat, repl, s)

In [18]:
len(s)

2102

We see the length has increased. That is the prrof of the replacement that has occured.

In [19]:
s

'In this chapter we will write a computer program implementing a neural network that learns to recognize handwritten digits. The program is just 74 lines long, and uses no special neural network libraries. But this short program can recognize digits with an accuracy over 96 percent, without human intervention. Furthermore, in later chapters we will develop ideas which can improve accuracy to over 99 percent. In fact, the best commercial neural networks are now so good that they are used by banks to process cheques, and by post offices to recognize addresses. We are focusing on handwriting recognition because it is an excellent prototype problem for learning about neural networks in general. As a prototype it hits a sweet spot: it is challenging - it is no small feat to recognize handwritten digits - but it is not so difficult as to require an extremely complicated solution, or tremendous computational power. Furthermore, it is a great way to develop more advanced techniques, such as de

Maybe this is a text from a news article so it will not have such words. 

In [20]:
def replaceWords(s):
    pattern = [(re.compile(regex), repl) for (regex, repl) in R_patterns ]
    for (pat, repl) in pattern:
        s = re.sub(pat, repl, s)

    return s

In [21]:
replaceWords("I won't do it.")

'I will not do it.'

In [22]:
replaceWords("Boy's quick.")

'Boy is quick.'

We can send multiple sentences to replace suck words.

In [23]:
sentences = sent_tokenize(raw_text)

In [24]:
len(sentences)

15

In [25]:
replaceWords(sentences[0])

'In this chapter we will write a computer program implementing a neural network that learns to recognize handwritten digits.'

In [26]:
from nltk.corpus import wordnet
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


True

In [27]:
def repeatingWord(word):
    repeat_regexp = re.compile(r'(\w*)(\w)\2(\w*)')
    repl = r'\1\2\3'
    if wordnet.synsets(word):
        return word
    repl_word = repeat_regexp.sub(repl, word)
    if repl_word != word :
        return repeatingWord(repl_word)
    else:
        return repl_word    

In [28]:
repeatingWord('Hiiiiiii')

'Hi'

In [29]:
repeatingWord('Hellooooo')

'Hello'

In [30]:
import pandas as pd

In [31]:
textFile = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/bag-of-words/vocab.nytimes.txt", sep='\n', header=None)

In [32]:
textFile.shape

(102660, 1)

In [33]:
textFile.head(10)

0
0         aah
1       aahed
2       aaron
3       aback
4      abacus
5       abajo
6     abalone
7     abandon
8   abandoned
9  abandoning

In [34]:
textFile[0]

0                            aah
1                          aahed
2                          aaron
3                          aback
4                         abacus
                   ...          
102655    zzz_zydrunas_ilgauskas
102656               zzz_zyprexa
102657                zzz_zyrtec
102658                zzz_zz_top
102659                  zzz_zzzt
Name: 0, Length: 102660, dtype: object

In [35]:
vocabList = textFile[0].values

In [36]:
vocabList

array(['aah', 'aahed', 'aaron', ..., 'zzz_zyrtec', 'zzz_zz_top',
       'zzz_zzzt'], dtype=object)

In [37]:
vocab = []
for w in vocabList:
    print(w)

Streaming output truncated to the last 5000 lines.
zzz_tom_kean
zzz_tom_kelley
zzz_tom_kelly
zzz_tom_king
zzz_tom_kite
zzz_tom_knight
zzz_tom_kuhn
zzz_tom_kuntz
zzz_tom_lampkin
zzz_tom_landry
zzz_tom_lasorda
zzz_tom_lea
zzz_tom_lee
zzz_tom_leedham
zzz_tom_lehman
zzz_tom_lehrer
zzz_tom_lemming
zzz_tom_levine
zzz_tom_liddy
zzz_tom_lieser
zzz_tom_lynch
zzz_tom_malchow
zzz_tom_malone
zzz_tom_malzbender
zzz_tom_martin
zzz_tom_mauser
zzz_tom_mccraw
zzz_tom_mcginnis
zzz_tom_meek
zzz_tom_menino
zzz_tom_miller
zzz_tom_minch
zzz_tom_modrak
zzz_tom_monaghan
zzz_tom_moore
zzz_tom_morello
zzz_tom_murphy
zzz_tom_mustin
zzz_tom_nalen
zzz_tom_noonan
zzz_tom_o_brien
zzz_tom_oder
zzz_tom_osborne
zzz_tom_parson
zzz_tom_pauken
zzz_tom_paulin
zzz_tom_payne
zzz_tom_pender
zzz_tom_perini
zzz_tom_peter
zzz_tom_petty
zzz_tom_phillip
zzz_tom_poti
zzz_tom_quann
zzz_tom_ramsey
zzz_tom_rath
zzz_tom_reilly
zzz_tom_rex
zzz_tom_ridge
zzz_tom_ripley
zzz_tom_robbin
zzz_tom_robinson
zzz_tom_roger
zzz_tom_rosenstiel
zzz_

In [38]:
"asdas".find("_")

-1

In [39]:
vocab = []
for w in range(len(vocabList)):
    if str(vocabList[w]).find('_') == -1:
        vocab.append(vocabList[w])

In [40]:
len(vocab)

44442

In [41]:
vocabSyn = []

In [42]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [43]:
from nltk.corpus import wordnet

In [44]:
wordnet.synsets('sun')[0].lemmas()[0].name

<bound method Lemma.name of Lemma('sun.n.01.sun')>

In [45]:
from google.colab import files

In [46]:
from google.colab import drive

In [47]:
drive.mount('/content/drive')

Mounted at /content/drive


In [48]:
filedf = pd.read_csv('/content/drive/My Drive/synonym.csv', header=None)

In [49]:
filedf

0         1
0  bday  Birthday
1    Hi     Hello
2    ok      Okay

In [50]:
x = [['ttyl', 'talk to you later'], ['lol', 'laughing out loud']]
x = pd.DataFrame(x)
filedf = pd.concat([filedf, x], ignore_index=True)
filedf

0                  1
0  bday           Birthday
1    Hi              Hello
2    ok               Okay
3  ttyl  talk to you later
4   lol  laughing out loud

In [51]:
import csv

class syn_replacer(object):

    def __init__(self):
        self.word_map = {}

        length = filedf.shape[0]
        for i in range(length):
            self.word_map[filedf.iloc[i,0]] = filedf.iloc[i,1]

    
    def replace(self, word):
        return self.word_map.get(word, word)

    def replace_sent(self, sent):
        w_list = sent.split(" ")
        w_list = [self.word_map.get(w, w) for w in w_list]
        return w_list

In [52]:
#synonym replacer's object
syn_rep = syn_replacer()
print(syn_rep.replace('ttyl'))
print(syn_rep.replace_sent('Today is my bday'))
print(syn_rep.replace_sent('lol hi, I will ttyl'))

talk to you later
['Today', 'is', 'my', 'Birthday']
['laughing out loud', 'hi,', 'I', 'will', 'talk to you later']


Antonyms

In [53]:
class word_antonym_replacer(object):

  def replace(self, word, pos=None):
    
    antonyms = set()
    
    for syn in wordnet.synsets(word, pos=pos):
        for lemma in syn.lemmas():
          for antonym in lemma.antonyms():
            antonyms.add(antonym.name())
    
    
    if len(antonyms) == 1:
        return antonyms.pop()
    else:
        return None
  
  def replace_negations(self, sent):
    i, l = 0, len(sent)
    words = []
    
    while i < l:
        word = sent[i]
        
        if word == 'not' and i+1 < l:
          ant = self.replace(sent[i+1])
          
          if ant:
              words.append(ant)
              i += 2
              continue
        
        words.append(word)
        i += 1
    
    return words

In [54]:
rep_ant = word_antonym_replacer()

print(rep_ant.replace('ugly'))
print(rep_ant.replace_negations(('This laptop is not heavy').split(' ')))
print(rep_ant.replace_negations(('Pyrotechniques are not dangerous').split(' ')))

beautiful
['This', 'laptop', 'is', 'light']
['Pyrotechniques', 'are', 'safe']
